### DISCUSSION TITLE
would a networking effect be a useful contribution?

### DISCUSSION THREAD

**Discussion by dark-ether** [haskell-effectful/effectful#143] [Ideas]
i am finally working on a project where i can use effectful and this made me realize that there isn't a networking effect ,i am working on it , would such an effect be acceptable in the effectful library? i am thinking of wrapping https://www.stackage.org/lts-20.11/package/socket-0.8.3.0

**Comment by ocharles** [comment]
Absolutely!  I'm not sure it needs to be part of effectful, but socket-effectful would be great

**Reply by arybczak** [reply]
Indeed, @dark-ether feel free to create a standalone library `socket-effectful` :+1:

**Comment by dark-ether** [comment]
@ocharles as you mentioned interest i have a question for you. would you prefer if the effect used bracket internally and stored the socket away from the user or do you prefer to have to manually do the error handling but have access to the socket?

**Reply by ocharles** [reply]
In all honesty - I don't know! This is exactly the dilemma I had when I opened https://github.com/haskell-effectful/effectful/discussions/137 :smile:

**Comment by Cajunvoodoo** [comment]
Apologies for necroposting, but I just hacked up a version for `network` (because of compatibility concerns, though `socket` would've been easy as well). It can be found at [github:Cajunvoodoo/network-effectful](https://github.com/Cajunvoodoo/network-effectful). I haven't had a chance to use it yet, and I'd need to look into whatever publishing on hackage entails, but hopefully this provides some momentum!

Regarding the "Should the effect store the socket" problem: I'm mixed on this. On one hand, `socket` is a wrapped function that produces a socket; how are you supposed to create a socket to run the socket effect (without IO)? Additionally, if the effect has no uniquely-determining type, only one socket would be usable at a time (unless a more complex value like a map is stored). The idea I had was something like `SocketEff 'SocketName`, but didn't bother going down that route for simplicity. It's an interesting idea, though.

**Reply by ocharles** [reply]
Effectful comes with labelled effects out-of-the-box. If you write an effect that scopes a single socket, you can easily give that a name. So no need for your socket effect to do naming

**Reply by arybczak** [reply]
I would imagine that if you're working with sockets, you will usually have multiple, so an effect per socket would be poor UX.
